openblas build fails using Cray compiler cce

`openblas@0.3.20` build fails using Cray compiler on OLCF Crusher.

I think this issue extends beyond Crusher, wherever Cray compilers are used to build openblas.

Spack compiler definition for cce:
```
compilers:
  - compiler:
      spec: cce@14.0.0
      paths:
        cc: /opt/cray/pe/craype/2.7.13/bin/cc
        cxx: /opt/cray/pe/craype/2.7.13/bin/CC
        f77: /opt/cray/pe/craype/2.7.13/bin/ftn
        fc: /opt/cray/pe/craype/2.7.13/bin/ftn
      flags: {}
      operating_system: sles15
      target: any
      modules:
      - PrgEnv-cray/8.3.3
      - cce/14.0.0
      - craype-x86-trento
      - libfabric
      - cray-pmi/6.1.2
```

Build error:
```
$> spack install openblas %cce@14
...
==> Installing openblas-0.3.20-wcx5q4myxg5skj63wzuf24bn2rhyfac4
==> No binary for openblas-0.3.20-wcx5q4myxg5skj63wzuf24bn2rhyfac4 found: installing from source
==> Fetching https://mirror.spack.io/_source-cache/archive/84/8495c9affc536253648e942908e88e097f2ec7753ede55aca52e5dead3029e3c.tar.gz
==> No patches needed for openblas
==> openblas: Executing phase: 'edit'
==> openblas: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16' 'CC=/gpfs/alpine/csc439/world-shared/E4S/ParaTools/22.05/PrgEnv-cray/spack/lib/spack/env/cce/cc' 'FC=/gpfs/alpine/csc439/world-shared/E4S/ParaTools/22.05/PrgEnv-cray/spack/lib/spack/env/cce/ftn' 'MAKE_NB_JOBS=0' 'ARCH=x86_64' 'TARGET=ZEN' 'USE_LOCKING=1' 'USE_OPENMP=1' 'USE_THREAD=1' 'RANLIB=ranlib' 'libs' 'netlib' 'shared'

18 errors found in build log:
     2879    ftn-2103 ftn: WARNING in command line
     2880      The -W all option is not supported or invalid and will be ignored.
     2881    ftn-2307 ftn: ERROR in command line
     2882      The "-m" option must be followed by 0, 1, 2, 3 or 4.
     2883    ftn-2307 ftn: ERROR in command line
     2884      The "-m" option must be followed by 0, 1, 2, 3 or 4.
  >> 2885    make[2]: *** [<builtin>: spotrf2.o] Error 1
     2886    make[2]: *** Waiting for unfinished jobs....
  >> 2887    make[2]: *** [<builtin>: sgecon.o] Error 1
  >> 2888    make[2]: *** [<builtin>: sgebak.o] Error 1
  >> 2889    make[2]: *** [<builtin>: sgbsvx.o] Error 1
  >> 2890    make[2]: *** [<builtin>: sgetrf2.o] Error 1
  >> 2891    make[2]: *** [<builtin>: sgbbrd.o] Error 1
  >> 2892    make[2]: *** [<builtin>: sgebal.o] Error 1
  >> 2893    make[2]: *** [<builtin>: sgbcon.o] Error 1
  >> 2894    make[2]: *** [<builtin>: sgbtrs.o] Error 1
  >> 2895    make[2]: *** [<builtin>: sgbrfs.o] Error 1
  >> 2896    make[2]: *** [<builtin>: sgebrd.o] Error 1
  >> 2897    make[2]: *** [<builtin>: sgebd2.o] Error 1
  >> 2898    make[2]: *** [<builtin>: sgbtrf.o] Error 1
  >> 2899    make[2]: *** [<builtin>: sgbequ.o] Error 1
  >> 2900    make[2]: *** [<builtin>: sgbtf2.o] Error 1
  >> 2901    make[2]: *** [<builtin>: sgbsv.o] Error 1
     2902    make[2]: Leaving directory '/tmp/eugenewalker/spack-stage/spack-stage-openblas-0.3.20-wcx5q4myxg5skj63wzuf24bn2rhyfac4/spack-src/lapack-netlib/SRC'
  >> 2903    make[1]: *** [Makefile:27: lapacklib] Error 2
     2904    make[1]: Leaving directory '/tmp/eugenewalker/spack-stage/spack-stage-openblas-0.3.20-wcx5q4myxg5skj63wzuf24bn2rhyfac4/spack-src/lapack-netlib'
  >> 2905    make: *** [Makefile:250: netlib] Error 2
```

FYI https://github.com/spack/spack/issues/30388

> Thank you. Let's try these replacements for f_check and Makefile.system: [cray.zip](https://github.com/xianyi/OpenBLAS/files/8929218/cray.zip)

Same looking build error:
* full log: [spack-build-out.txt](https://github.com/xianyi/OpenBLAS/files/8929401/spack-build-out.txt)
 
```
==> openblas: Executing phase: 'edit'
==> openblas: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16' 'CC=/gpfs/alpine/csc439/world-shared/E4S/ParaTools/22.05/PrgEnv-cray/spack/lib/spack/env/cce/cc' 'FC=/gpfs/alpine/csc439/world-shared/E4S/ParaTools/22.05/PrgEnv-cray/spack/lib/spack/env/cce/ftn' 'MAKE_NB_JOBS=0' 'ARCH=x86_64' 'TARGET=ZEN' 'USE_LOCKING=1' 'USE_OPENMP=1' 'USE_THREAD=1' 'RANLIB=ranlib' 'libs' 'netlib' 'shared'

18 errors found in build log:
     132    ftn-2307 ftn: ERROR in command line
     133      The "-m" option must be followed by 0, 1, 2, 3 or 4.
     134    ftn-2307 ftn: ERROR in command line
     135      The "-m" option must be followed by 0, 1, 2, 3 or 4.
     136    ftn-2307 ftn: ERROR in command line
     137      The "-m" option must be followed by 0, 1, 2, 3 or 4.
  >> 138    make[2]: *** [<builtin>: sgbbrd.o] Error 1
     139    make[2]: *** Waiting for unfinished jobs....
  >> 140    make[2]: *** [<builtin>: sgetrf2.o] Error 1
  >> 141    make[2]: *** [<builtin>: sgbtrf.o] Error 1
  >> 142    make[2]: *** [<builtin>: sgebal.o] Error 1
  >> 143    make[2]: *** [<builtin>: sgbequ.o] Error 1
  >> 144    make[2]: *** [<builtin>: sgebrd.o] Error 1
  >> 145    make[2]: *** [<builtin>: sgebd2.o] Error 1
  >> 146    make[2]: *** [<builtin>: spotrf2.o] Error 1
  >> 147    make[2]: *** [<builtin>: sgbcon.o] Error 1
  >> 148    make[2]: *** [<builtin>: sgebak.o] Error 1
  >> 149    make[2]: *** [<builtin>: sgbsvx.o] Error 1
  >> 150    make[2]: *** [<builtin>: sgbtrs.o] Error 1
  >> 151    make[2]: *** [<builtin>: sgbsv.o] Error 1
  >> 152    make[2]: *** [<builtin>: sgecon.o] Error 1
  >> 153    make[2]: *** [<builtin>: sgbrfs.o] Error 1
  >> 154    make[2]: *** [<builtin>: sgbtf2.o] Error 1
     155    make[2]: Leaving directory '/tmp/lpeyrala/spack-stage/spack-stage-openblas-0.3.20-wcx5q4myxg5skj63wzuf24bn2rhyfac4/spack-src/lapack-netlib/SRC'
  >> 156    make[1]: *** [Makefile:27: lapacklib] Error 2
     157    make[1]: Leaving directory '/tmp/lpeyrala/spack-stage/spack-stage-openblas-0.3.20-wcx5q4myxg5skj63wzuf24bn2rhyfac4/spack-src/lapack-netlib'
  >> 158    make: *** [Makefile:250: netlib] Error 2
```

> Ah silly me - that is the CFLAGS spilling over into the Fortran part of the build. We already sanitize these for the NAG compiler so this will hopefully get us to some new error message [Makefile.system.txt](https://github.com/xianyi/OpenBLAS/files/8929481/Makefile.system.txt)

Next up:
* Full log: [spack-build-out.txt](https://github.com/xianyi/OpenBLAS/files/8929548/spack-build-out.txt)

Abbreviated:
```
==> openblas: Executing phase: 'edit'
==> openblas: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16' 'CC=/gpfs/alpine/csc439/world-shared/E4S/ParaTools/22.05/PrgEnv-cray/spack/lib/spack/env/cce/cc' 'FC=/gpfs/alpine/csc439/world-shared/E4S/ParaTools/22.05/PrgEnv-cray/spack/lib/spack/env/cce/ftn' 'MAKE_NB_JOBS=0' 'ARCH=x86_64' 'TARGET=ZEN' 'USE_LOCKING=1' 'USE_OPENMP=1' 'USE_THREAD=1' 'RANLIB=ranlib' 'libs' 'netlib' 'shared'

7 errors found in build log:
     4631    make[1]: Entering directory '/tmp/lpeyrala/spack-stage/spack-stage-openblas-0.3.20-wcx5q4myxg5skj63wzuf24bn2rhyfac4/spack-src/exports'
     4632    perl ./gensymbol linktest  x86_64 _ 0 0 0 0 0 0 "" "" 1 0 1 1 1 1 > linktest.c
     4633    /gpfs/alpine/csc439/world-shared/E4S/ParaTools/22.05/PrgEnv-cray/spack/lib/spack/env/cce/cc -O2 -DSMALL_MATRIX_OPT -DMAX_STACK_ALLOC=2048 -DUSE_LOCKING -fopenmp -Wall -m64 -fPIC -DSMP
             _SERVER -DUSE_OPENMP -DNO_WARMUP -DMAX_CPU_NUMBER=128 -DMAX_PARALLEL_NUMBER=1 -DBUILD_SINGLE=1 -DBUILD_DOUBLE=1 -DBUILD_COMPLEX=1 -DBUILD_COMPLEX16=1 -DVERSION=\"0.3.20\" -msse3 -msse
             4.1 -mavx -mavx2 -mavx2 -UASMNAME -UASMFNAME -UNAME -UCNAME -UCHAR_NAME -UCHAR_CNAME -DASMNAME= -DASMFNAME=_ -DNAME=_ -DCNAME= -DCHAR_NAME=\"_\" -DCHAR_CNAME=\"\" -DNO_AFFINITY -I..
             -shared -o ../libopenblas_zenp-r0.3.20.so \
     4634    -Wl,--whole-archive ../libopenblas_zenp-r0.3.20.a -Wl,--no-whole-archive \
     4635    -Wl,-soname,libopenblas.so.0 -lm -lpthread -lm -lpthread
     4636    /gpfs/alpine/csc439/world-shared/E4S/ParaTools/22.05/PrgEnv-cray/spack/lib/spack/env/cce/cc -O2 -DSMALL_MATRIX_OPT -DMAX_STACK_ALLOC=2048 -DUSE_LOCKING -fopenmp -Wall -m64 -fPIC -DSMP
             _SERVER -DUSE_OPENMP -DNO_WARMUP -DMAX_CPU_NUMBER=128 -DMAX_PARALLEL_NUMBER=1 -DBUILD_SINGLE=1 -DBUILD_DOUBLE=1 -DBUILD_COMPLEX=1 -DBUILD_COMPLEX16=1 -DVERSION=\"0.3.20\" -msse3 -msse
             4.1 -mavx -mavx2 -mavx2 -UASMNAME -UASMFNAME -UNAME -UCNAME -UCHAR_NAME -UCHAR_CNAME -DASMNAME= -DASMFNAME=_ -DNAME=_ -DCNAME= -DCHAR_NAME=\"_\" -DCHAR_CNAME=\"\" -DNO_AFFINITY -I..
             -w -o linktest linktest.c ../libopenblas_zenp-r0.3.20.so -L/opt/cray/pe/gcc/10.3.0/snos/lib64 -L/opt/cray/pe/pmi/6.1.2/lib -L/opt/cray/pe/dsmml/0.2.2/dsmml//lib -L/opt/cray/pe/cce/14.
             0.0/cce/x86_64/lib/pkgconfig/../ -L/opt/gcc/10.3.0/snos/lib64 -L/opt/cray/pe/gcc/10.3.0/snos/lib64 -L/opt/cray/pe/cce/14.0.0/cce/x86_64/lib/pkgconfig/../ -L/opt/cray/pe/cce/14.0.0/cce
             -clang/x86_64/lib/clang/14.0.0/lib/linux -L/opt/cray/pe/gcc/10.3.0/snos/lib/gcc/x86_64-suse-linux/10.3.0 -L/opt/cray/pe/cce/14.0.0/binutils/x86_64/x86_64-pc-linux-gnu/..//x86_64-unkno
             wn-linux-gnu/lib  -lpmi -ldsmml -lstdc++ -lpgas-shmem -lquadmath -lcraymp -lmodules -lfi -lcraymath -lf -lu -lcsup -latomic -lgfortran -lstdc++ -lpthread -lc -lcsup -lm  -usystem_mod=
             /opt/cray/pe/cce/14.0.0/cce/x86_64/lib/libmodules.a -usystem_mod=/opt/cray/pe/cce/14.0.0/cce/x86_64/lib/libomp.a -usystem_mod=/opt/cray/pe/cce/14.0.0/cce/x86_64/lib/omp_lib.a -usystem
             _mod=/opt/cray/pe/cce/14.0.0/cce/x86_64/lib/libopenacc.a -usystem_mod=/opt/cray/pe/cce/14.0.0/cce/x86_64/lib/libcraymp.a  && echo OK.
  >> 4637    ld.lld: error: ../libopenblas_zenp-r0.3.20.so: undefined reference to _ismin_ [--no-allow-shlib-undefined]
  >> 4638    ld.lld: error: ../libopenblas_zenp-r0.3.20.so: undefined reference to _idmin_ [--no-allow-shlib-undefined]
  >> 4639    ld.lld: error: ../libopenblas_zenp-r0.3.20.so: undefined reference to _ismax_ [--no-allow-shlib-undefined]
  >> 4640    ld.lld: error: ../libopenblas_zenp-r0.3.20.so: undefined reference to _idmax_ [--no-allow-shlib-undefined]
  >> 4641    clang-14: error: linker command failed with exit code 1 (use -v to see invocation)
  >> 4642    make[1]: *** [Makefile:190: ../libopenblas_zenp-r0.3.20.so] Error 1
     4643    make[1]: Leaving directory '/tmp/lpeyrala/spack-stage/spack-stage-openblas-0.3.20-wcx5q4myxg5skj63wzuf24bn2rhyfac4/spack-src/exports'
  >> 4644    make: *** [Makefile:122: shared] Error 2
```

> > Instead of using `CC/cc/ftn`, can you try `crayCC/craycc/crayftn`?
> 
> In that case, I will add `crayftn` as well to the list of recognized fortran compilers in f_check -- is that right?

I tried using the following patched files:
* `Makefile.system`
* `f_check` (modified to additionally recognize `crayftn`)

With this Spack compiler definition:
```
compilers:
  - compiler:
      spec: cce@14.0.0
      paths:
        cc: craycc
        cxx: crayCC
        f77: crayftn
        fc: crayftn
      flags: {}
      operating_system: sles15
      target: any
      modules:
      - PrgEnv-cray/8.3.3
      - cce/14.0.0
      - craype-x86-trento
      - libfabric
      - cray-pmi/6.1.2
```

And I still got this error:
```
==> openblas: Executing phase: 'edit'
==> openblas: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16' 'CC=/autofs/nccs-svm1_home1/lpeyrala/openblas-cce/spack/lib/spack/env/cce/cc' 'FC=/autofs/nccs-svm1_home1/lpeyrala/openblas-cce/spack/lib/spack/env/cce/ftn' 'MAKE_NB_JOBS=0' 'ARCH=x86_64' 'TARGET=ZEN' 'USE_LOCKING=1' 'USE_OPENMP=0' 'USE_THREAD=0' 'RANLIB=ranlib' 'libs' 'netlib' 'shared'

7 errors found in build log:
     6566    make[1]: Entering directory '/tmp/lpeyrala/spack-stage/spack-stage-openblas-0.3.20-sdak36kpmvoemo3fnfpjqvgykboog733/spack-src/exports'
     6567    perl ./gensymbol linktest  x86_64 _ 0 0 0 0 0 0 "" "" 1 0 1 1 1 1 > linktest.c
     6568    /autofs/nccs-svm1_home1/lpeyrala/openblas-cce/spack/lib/spack/env/cce/cc -O2 -DSMALL_MATRIX_OPT -DMAX_STACK_ALLOC=2048 -DUSE_LOCKING -Wall -m64 -fPIC -DNO_WARMUP -DMAX_CPU_NUMBER=128
             -DMAX_PARALLEL_NUMBER=1 -DBUILD_SINGLE=1 -DBUILD_DOUBLE=1 -DBUILD_COMPLEX=1 -DBUILD_COMPLEX16=1 -DVERSION=\"0.3.20\" -msse3 -msse4.1 -mavx -mavx2 -mavx2 -UASMNAME -UASMFNAME -UNAME -U
             CNAME -UCHAR_NAME -UCHAR_CNAME -DASMNAME= -DASMFNAME=_ -DNAME=_ -DCNAME= -DCHAR_NAME=\"_\" -DCHAR_CNAME=\"\" -DNO_AFFINITY -I..  -shared -o ../libopenblas_zen-r0.3.20.so \
     6569    -Wl,--whole-archive ../libopenblas_zen-r0.3.20.a -Wl,--no-whole-archive \
     6570    -Wl,-soname,libopenblas.so.0 -lm -lm
     6571    /autofs/nccs-svm1_home1/lpeyrala/openblas-cce/spack/lib/spack/env/cce/cc -O2 -DSMALL_MATRIX_OPT -DMAX_STACK_ALLOC=2048 -DUSE_LOCKING -Wall -m64 -fPIC -DNO_WARMUP -DMAX_CPU_NUMBER=128
             -DMAX_PARALLEL_NUMBER=1 -DBUILD_SINGLE=1 -DBUILD_DOUBLE=1 -DBUILD_COMPLEX=1 -DBUILD_COMPLEX16=1 -DVERSION=\"0.3.20\" -msse3 -msse4.1 -mavx -mavx2 -mavx2 -UASMNAME -UASMFNAME -UNAME -U
             CNAME -UCHAR_NAME -UCHAR_CNAME -DASMNAME= -DASMFNAME=_ -DNAME=_ -DCNAME= -DCHAR_NAME=\"_\" -DCHAR_CNAME=\"\" -DNO_AFFINITY -I..  -w -o linktest linktest.c ../libopenblas_zen-r0.3.20.s
             o -L/opt/gcc/8.1.0/snos/lib64 -L/opt/gcc/10.3.0/snos/lib64 -L/opt/cray/pe/gcc/10.3.0/snos/lib64 -L/opt/cray/pe/cce/14.0.0/cce/x86_64/lib/pkgconfig/../ -L/opt/cray/pe/cce/14.0.0/cce-cl
             ang/x86_64/lib/clang/14.0.0/lib/linux -L/opt/gcc/8.1.0/snos/lib/gcc/x86_64-suse-linux/8.1.0 -L/opt/cray/pe/cce/14.0.0/binutils/x86_64/x86_64-pc-linux-gnu/..//x86_64-unknown-linux-gnu/
             lib  -lquadmath -lmodules -lfi -lcraymath -lf -lu -lcsup -lgfortran -ltcmalloc_minimal -lstdc++ -lpthread -lc -lcsup -lm  -usystem_mod=/opt/cray/pe/cce/14.0.0/cce/x86_64/lib/libmodule
             s.a -usystem_mod=/opt/cray/pe/cce/14.0.0/cce/x86_64/lib/libomp.a -usystem_mod=/opt/cray/pe/cce/14.0.0/cce/x86_64/lib/omp_lib.a -usystem_mod=/opt/cray/pe/cce/14.0.0/cce/x86_64/lib/libo
             penacc.a -usystem_mod=/opt/cray/pe/cce/14.0.0/cce/x86_64/lib/libcraymp.a  && echo OK.
  >> 6572    ld.lld: error: ../libopenblas_zen-r0.3.20.so: undefined reference to _ismin_ [--no-allow-shlib-undefined]
  >> 6573    ld.lld: error: ../libopenblas_zen-r0.3.20.so: undefined reference to _idmin_ [--no-allow-shlib-undefined]
  >> 6574    ld.lld: error: ../libopenblas_zen-r0.3.20.so: undefined reference to _ismax_ [--no-allow-shlib-undefined]
  >> 6575    ld.lld: error: ../libopenblas_zen-r0.3.20.so: undefined reference to _idmax_ [--no-allow-shlib-undefined]
  >> 6576    clang-14: error: linker command failed with exit code 1 (use -v to see invocation)
  >> 6577    make[1]: *** [Makefile:190: ../libopenblas_zen-r0.3.20.so] Error 1
     6578    make[1]: Leaving directory '/tmp/lpeyrala/spack-stage/spack-stage-openblas-0.3.20-sdak36kpmvoemo3fnfpjqvgykboog733/spack-src/exports'
  >> 6579    make: *** [Makefile:122: shared] Error 2
```

I'm not sure how these references are getting in there now.

@lukebroskop any ideas on next steps? I can dig in further I'm just temporarily stalled as to how to proceed.

> Can you try with the replacement KERNEL.ZEN file please ?

Full log: [spack-build-out.txt](https://github.com/xianyi/OpenBLAS/files/8931546/spack-build-out.txt)

Same:
```
==> openblas: Executing phase: 'edit'
==> openblas: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16' 'CC=/autofs/nccs-svm1_home1/lpeyrala/openblas-cce/spack/lib/spack/env/cce/cc' 'FC=/autofs/nccs-svm1_home1/lpeyrala/openblas-cce/spack/lib/spack/env/cce/ftn' 'MAKE_NB_JOBS=0' 'ARCH=x86_64' 'TARGET=ZEN' 'USE_LOCKING=1' 'USE_OPENMP=0' 'USE_THREAD=0' 'RANLIB=ranlib' 'libs' 'netlib' 'shared'

7 errors found in build log:
     6566    make[1]: Entering directory '/tmp/lpeyrala/spack-stage/spack-stage-openblas-0.3.20-sdak36kpmvoemo3fnfpjqvgykboog733/spack-src/exports'
     6567    perl ./gensymbol linktest  x86_64 _ 0 0 0 0 0 0 "" "" 1 0 1 1 1 1 > linktest.c
     6568    /autofs/nccs-svm1_home1/lpeyrala/openblas-cce/spack/lib/spack/env/cce/cc -O2 -DSMALL_MATRIX_OPT -DMAX_STACK_ALLOC=2048 -DUSE_LOCKING -Wall -m64 -fPIC -DNO_WARMUP -DMAX_CPU_NUMBER=128
             -DMAX_PARALLEL_NUMBER=1 -DBUILD_SINGLE=1 -DBUILD_DOUBLE=1 -DBUILD_COMPLEX=1 -DBUILD_COMPLEX16=1 -DVERSION=\"0.3.20\" -msse3 -msse4.1 -mavx -mavx2 -mavx2 -UASMNAME -UASMFNAME -UNAME -U
             CNAME -UCHAR_NAME -UCHAR_CNAME -DASMNAME= -DASMFNAME=_ -DNAME=_ -DCNAME= -DCHAR_NAME=\"_\" -DCHAR_CNAME=\"\" -DNO_AFFINITY -I..  -shared -o ../libopenblas_zen-r0.3.20.so \
     6569    -Wl,--whole-archive ../libopenblas_zen-r0.3.20.a -Wl,--no-whole-archive \
     6570    -Wl,-soname,libopenblas.so.0 -lm -lm
     6571    /autofs/nccs-svm1_home1/lpeyrala/openblas-cce/spack/lib/spack/env/cce/cc -O2 -DSMALL_MATRIX_OPT -DMAX_STACK_ALLOC=2048 -DUSE_LOCKING -Wall -m64 -fPIC -DNO_WARMUP -DMAX_CPU_NUMBER=128
             -DMAX_PARALLEL_NUMBER=1 -DBUILD_SINGLE=1 -DBUILD_DOUBLE=1 -DBUILD_COMPLEX=1 -DBUILD_COMPLEX16=1 -DVERSION=\"0.3.20\" -msse3 -msse4.1 -mavx -mavx2 -mavx2 -UASMNAME -UASMFNAME -UNAME -U
             CNAME -UCHAR_NAME -UCHAR_CNAME -DASMNAME= -DASMFNAME=_ -DNAME=_ -DCNAME= -DCHAR_NAME=\"_\" -DCHAR_CNAME=\"\" -DNO_AFFINITY -I..  -w -o linktest linktest.c ../libopenblas_zen-r0.3.20.s
             o -L/opt/gcc/8.1.0/snos/lib64 -L/opt/gcc/10.3.0/snos/lib64 -L/opt/cray/pe/gcc/10.3.0/snos/lib64 -L/opt/cray/pe/cce/14.0.0/cce/x86_64/lib/pkgconfig/../ -L/opt/cray/pe/cce/14.0.0/cce-cl
             ang/x86_64/lib/clang/14.0.0/lib/linux -L/opt/gcc/8.1.0/snos/lib/gcc/x86_64-suse-linux/8.1.0 -L/opt/cray/pe/cce/14.0.0/binutils/x86_64/x86_64-pc-linux-gnu/..//x86_64-unknown-linux-gnu/
             lib  -lquadmath -lmodules -lfi -lcraymath -lf -lu -lcsup -lgfortran -ltcmalloc_minimal -lstdc++ -lpthread -lc -lcsup -lm  -usystem_mod=/opt/cray/pe/cce/14.0.0/cce/x86_64/lib/libmodule
             s.a -usystem_mod=/opt/cray/pe/cce/14.0.0/cce/x86_64/lib/libomp.a -usystem_mod=/opt/cray/pe/cce/14.0.0/cce/x86_64/lib/omp_lib.a -usystem_mod=/opt/cray/pe/cce/14.0.0/cce/x86_64/lib/libo
             penacc.a -usystem_mod=/opt/cray/pe/cce/14.0.0/cce/x86_64/lib/libcraymp.a  && echo OK.
  >> 6572    ld.lld: error: ../libopenblas_zen-r0.3.20.so: undefined reference to _ismin_ [--no-allow-shlib-undefined]
  >> 6573    ld.lld: error: ../libopenblas_zen-r0.3.20.so: undefined reference to _idmin_ [--no-allow-shlib-undefined]
  >> 6574    ld.lld: error: ../libopenblas_zen-r0.3.20.so: undefined reference to _ismax_ [--no-allow-shlib-undefined]
  >> 6575    ld.lld: error: ../libopenblas_zen-r0.3.20.so: undefined reference to _idmax_ [--no-allow-shlib-undefined]
  >> 6576    clang-14: error: linker command failed with exit code 1 (use -v to see invocation)
  >> 6577    make[1]: *** [Makefile:190: ../libopenblas_zen-r0.3.20.so] Error 1
     6578    make[1]: Leaving directory '/tmp/lpeyrala/spack-stage/spack-stage-openblas-0.3.20-sdak36kpmvoemo3fnfpjqvgykboog733/spack-src/exports'
  >> 6579    make: *** [Makefile:122: shared] Error 2
```

> > Are you sure that your spack magic maps your FC of `/autofs/nccs-svm1_home1/lpeyrala/openblas-cce/spack/lib/spack/env/cce/ftn` to "crayftn" rather than the wrapper ? (Probably a silly question, but I am running out of ideas). Perhaps upload the generated libopenblas.a (or check if there are more symbols with prepended underscores in it besides the four that are giving us grief)
> 
> I will look for confirmation this is indeed happening the way it should.

Spack is invoking `make` like this:
```
make -j16 \
 CC=<spack-compiler-wrapper-cc> \
 FC=<spack-compiler-wrapper-ftn> \
 MAKE_NB_JOBS=0 \
 ARCH=x86_64 \
 TARGET=ZEN \
 USE_LOCKING=1 \
 USE_OPENMP=0 \
 USE_THREAD=0 \
 RANLIB=ranlib \
 libs netlib shared
```

I did:
1. `make clean`
2. re-invoked `make` the same as above except substituting `craycc/crayftn` in place of the Spack wrappers.

And I got the same error:

```
$> make clean
$> make ...
...
craycc -O2 -DSMALL_MATRIX_OPT -DMAX_STACK_ALLOC=2048 -DUSE_LOCKING -Wall -m64 -fPIC -DNO_WARMUP -DMAX_CPU_NUMBER=128 -DMAX_PARALLEL_NUMBER=1 -DBUILD_SINGLE=1 -DBUILD_DOUBLE=1 -DBUILD_COMPLEX=1 -DBUILD_COMPLEX16=1 -DVERSION=\"0.3.20\" -msse3 -msse4.1 -mavx -mavx2 -mavx2 -UASMNAME -UASMFNAME -UNAME -UCNAME -UCHAR_NAME -UCHAR_CNAME -DASMNAME= -DASMFNAME=_ -DNAME=_ -DCNAME= -DCHAR_NAME=\"_\" -DCHAR_CNAME=\"\" -DNO_AFFINITY -I..  -shared -o ../libopenblas_zen-r0.3.20.so \
-Wl,--whole-archive ../libopenblas_zen-r0.3.20.a -Wl,--no-whole-archive \
-Wl,-soname,libopenblas.so.0 -lm -lm
craycc -O2 -DSMALL_MATRIX_OPT -DMAX_STACK_ALLOC=2048 -DUSE_LOCKING -Wall -m64 -fPIC -DNO_WARMUP -DMAX_CPU_NUMBER=128 -DMAX_PARALLEL_NUMBER=1 -DBUILD_SINGLE=1 -DBUILD_DOUBLE=1 -DBUILD_COMPLEX=1 -DBUILD_COMPLEX16=1 -DVERSION=\"0.3.20\" -msse3 -msse4.1 -mavx -mavx2 -mavx2 -UASMNAME -UASMFNAME -UNAME -UCNAME -UCHAR_NAME -UCHAR_CNAME -DASMNAME= -DASMFNAME=_ -DNAME=_ -DCNAME= -DCHAR_NAME=\"_\" -DCHAR_CNAME=\"\" -DNO_AFFINITY -I..  -w -o linktest linktest.c ../libopenblas_zen-r0.3.20.so -L/opt/gcc/8.1.0/snos/lib64 -L/opt/gcc/10.3.0/snos/lib64 -L/opt/cray/pe/gcc/10.3.0/snos/lib64 -L/opt/cray/pe/cce/14.0.0/cce/x86_64/lib/pkgconfig/../ -L/opt/cray/pe/cce/14.0.0/cce-clang/x86_64/lib/clang/14.0.0/lib/linux -L/opt/gcc/8.1.0/snos/lib/gcc/x86_64-suse-linux/8.1.0 -L/opt/cray/pe/cce/14.0.0/binutils/x86_64/x86_64-pc-linux-gnu/..//x86_64-unknown-linux-gnu/lib  -lquadmath -lmodules -lfi -lcraymath -lf -lu -lcsup -lgfortran -ltcmalloc_minimal -lstdc++ -lpthread -lc -lcsup -lm  -usystem_mod=/opt/cray/pe/cce/14.0.0/cce/x86_64/lib/libmodules.a -usystem_mod=/opt/cray/pe/cce/14.0.0/cce/x86_64/lib/libomp.a -usystem_mod=/opt/cray/pe/cce/14.0.0/cce/x86_64/lib/omp_lib.a -usystem_mod=/opt/cray/pe/cce/14.0.0/cce/x86_64/lib/libopenacc.a -usystem_mod=/opt/cray/pe/cce/14.0.0/cce/x86_64/lib/libcraymp.a  && echo OK.
ld.lld: error: ../libopenblas_zen-r0.3.20.so: undefined reference to _ismin_ [--no-allow-shlib-undefined]
ld.lld: error: ../libopenblas_zen-r0.3.20.so: undefined reference to _idmin_ [--no-allow-shlib-undefined]
ld.lld: error: ../libopenblas_zen-r0.3.20.so: undefined reference to _ismax_ [--no-allow-shlib-undefined]
ld.lld: error: ../libopenblas_zen-r0.3.20.so: undefined reference to _idmax_ [--no-allow-shlib-undefined]
clang-14: error: linker command failed with exit code 1 (use -v to see invocation)
make[1]: *** [Makefile:190: ../libopenblas_zen-r0.3.20.so] Error 1
make[1]: Leaving directory '/tmp/lpeyrala/spack-stage/spack-stage-openblas-0.3.20-sdak36kpmvoemo3fnfpjqvgykboog733/spack-src/exports'
make: *** [Makefile:122: shared] Error 2
```

> Just do a `make clean` and `make` when you made the change - this should simply force leading underscores on the C symbol names to (hopefully) make the naming consistent between the two compilers so that the linking should succeed.

In `c_check` I changed
```
$need_fu      = $1;
```
to
```
$need_fu      = 1;
```

And then:
```
$> make -j16 \
 CC=craycc \
 FC=crayftn \
 MAKE_NB_JOBS=0 \
 ARCH=x86_64 \
 TARGET=ZEN \
 USE_LOCKING=1 \
 USE_OPENMP=0 \
 USE_THREAD=0 \
 RANLIB=ranlib \
 libs netlib shared
...
../kernel/x86_64/asum_sse.S:50:28: error: expected identifier
 .text; .align 512; .globl 1sasum_k ; .type 1sasum_k, @function; 1sasum_k:
                           ^
../kernel/x86_64/asum_sse.S:50:45: error: expected identifier in directive
 .text; .align 512; .globl 1sasum_k ; .type 1sasum_k, @function; 1sasum_k:
                                            ^
../kernel/x86_64/asum_sse.S:50:66: error: unexpected token at start of statement
 .text; .align 512; .globl 1sasum_k ; .type 1sasum_k, @function; 1sasum_k:
                                                                 ^
../kernel/x86_64/asum_sse.S:345:8: error: expected identifier in directive
 .size 1sasum_k, .-1sasum_k; .section .note.GNU-stack,"",@progbits
       ^
make[1]: *** [Makefile.L1:680: sasum_k.o] Error 1
../kernel/x86_64/nrm2_sse.S:50:28: error: expected identifier
 .text; .align 512; .globl 1snrm2_k ; .type 1snrm2_k, @function; 1snrm2_k:
                           ^
../kernel/x86_64/nrm2_sse.S:50:45: error: expected identifier in directive
 .text; .align 512; .globl 1snrm2_k ; .type 1snrm2_k, @function; 1snrm2_k:
                                            ^
../kernel/x86_64/nrm2_sse.S:50:66: error: unexpected token at start of statement
 .text; .align 512; .globl 1snrm2_k ; .type 1snrm2_k, @function; 1snrm2_k:
                                                                 ^
../kernel/x86_64/nrm2_sse.S:315:8: error: expected identifier in directive
 .size 1snrm2_k, .-1snrm2_k; .section .note.GNU-stack,"",@progbits
       ^
make[1]: *** [Makefile.L1:821: snrm2_k.o] Error 1
../kernel/x86_64/copy_sse.S:60:28: error: expected identifier
 .text; .align 512; .globl 1scopy_k ; .type 1scopy_k, @function; 1scopy_k:
                           ^
../kernel/x86_64/copy_sse.S:60:45: error: expected identifier in directive
 .text; .align 512; .globl 1scopy_k ; .type 1scopy_k, @function; 1scopy_k:
                                            ^
../kernel/x86_64/copy_sse.S:60:66: error: unexpected token at start of statement
 .text; .align 512; .globl 1scopy_k ; .type 1scopy_k, @function; 1scopy_k:
                                                                 ^
../kernel/x86_64/copy_sse.S:959:8: error: expected identifier in directive
 .size 1scopy_k, .-1scopy_k; .section .note.GNU-stack,"",@progbits
       ^
make[1]: *** [Makefile.L1:745: scopy_k.o] Error 1
make[1]: Leaving directory '/tmp/lpeyrala/spack-stage/spack-stage-openblas-0.3.20-sdak36kpmvoemo3fnfpjqvgykboog733/spack-src/kernel'
make: *** [Makefile:170: libs] Error 1
```

> Ugh. This sort-of suggests that it would be better to change that line to `$need_fu = _` , sorry for misreading the perl code.

```
$> make...
...
craycc -O2 -DSMALL_MATRIX_OPT -DMAX_STACK_ALLOC=2048 -DUSE_LOCKING -Wall -m64 -fPIC -DNO_WARMUP -DMAX_CPU_NUMBER=128 -DMAX_PARALLEL_NUMBER=1 -DBUILD_SINGLE=1 -DBUILD_DOUBLE=1 -DBUILD_COMPLEX=1 -DBUILD_COMPLEX16=1 -DVERSION=\"0.3.20\" -msse3 -msse4.1 -mavx -mavx2 -mavx2 -UASMNAME -UASMFNAME -UNAME -UCNAME -UCHAR_NAME -UCHAR_CNAME -DASMNAME=_ -DASMFNAME=__ -DNAME=_ -DCNAME= -DCHAR_NAME=\"_\" -DCHAR_CNAME=\"\" -DNO_AFFINITY -I..  -w -o linktest linktest.c ../libopenblas_zen-r0.3.20.so -L/opt/gcc/8.1.0/snos/lib64 -L/opt/gcc/10.3.0/snos/lib64 -L/opt/cray/pe/gcc/10.3.0/snos/lib64 -L/opt/cray/pe/cce/14.0.0/cce/x86_64/lib/pkgconfig/../ -L/opt/cray/pe/cce/14.0.0/cce-clang/x86_64/lib/clang/14.0.0/lib/linux -L/opt/gcc/8.1.0/snos/lib/gcc/x86_64-suse-linux/8.1.0 -L/opt/cray/pe/cce/14.0.0/binutils/x86_64/x86_64-pc-linux-gnu/..//x86_64-unknown-linux-gnu/lib  -lquadmath -lmodules -lfi -lcraymath -lf -lu -lcsup -lgfortran -ltcmalloc_minimal -lstdc++ -lpthread -lc -lcsup -lm  -usystem_mod=/opt/cray/pe/cce/14.0.0/cce/x86_64/lib/libmodules.a -usystem_mod=/opt/cray/pe/cce/14.0.0/cce/x86_64/lib/libomp.a -usystem_mod=/opt/cray/pe/cce/14.0.0/cce/x86_64/lib/omp_lib.a -usystem_mod=/opt/cray/pe/cce/14.0.0/cce/x86_64/lib/libopenacc.a -usystem_mod=/opt/cray/pe/cce/14.0.0/cce/x86_64/lib/libcraymp.a  && echo OK.
ld.lld: error: undefined symbol: lsame_
>>> referenced by linktest.c
>>>               /tmp/linktest-c81c97.o:(main)
>>> did you mean: _lsame_
>>> defined in: ../libopenblas_zen-r0.3.20.so

ld.lld: error: undefined symbol: dcabs1_
>>> referenced by linktest.c
>>>               /tmp/linktest-c81c97.o:(main)
>>> did you mean: _dcabs1_
>>> defined in: ../libopenblas_zen-r0.3.20.so

ld.lld: error: undefined symbol: scabs1_
>>> referenced by linktest.c
>>>               /tmp/linktest-c81c97.o:(main)

ld.lld: error: ../libopenblas_zen-r0.3.20.so: undefined reference to sswap_k [--no-allow-shlib-undefined]
ld.lld: error: ../libopenblas_zen-r0.3.20.so: undefined reference to scopy_k [--no-allow-shlib-undefined]
ld.lld: error: ../libopenblas_zen-r0.3.20.so: undefined reference to sscal_k [--no-allow-shlib-undefined]
ld.lld: error: ../libopenblas_zen-r0.3.20.so: undefined reference to sasum_k [--no-allow-shlib-undefined]
ld.lld: error: ../libopenblas_zen-r0.3.20.so: undefined reference to snrm2_k [--no-allow-shlib-undefined]
ld.lld: error: ../libopenblas_zen-r0.3.20.so: undefined reference to smax_k [--no-allow-shlib-undefined]
ld.lld: error: ../libopenblas_zen-r0.3.20.so: undefined reference to samax_k [--no-allow-shlib-undefined]
ld.lld: error: ../libopenblas_zen-r0.3.20.so: undefined reference to isamax_k [--no-allow-shlib-undefined]
ld.lld: error: ../libopenblas_zen-r0.3.20.so: undefined reference to smin_k [--no-allow-shlib-undefined]
ld.lld: error: ../libopenblas_zen-r0.3.20.so: undefined reference to samin_k [--no-allow-shlib-undefined]
ld.lld: error: ../libopenblas_zen-r0.3.20.so: undefined reference to isamin_k [--no-allow-shlib-undefined]
ld.lld: error: ../libopenblas_zen-r0.3.20.so: undefined reference to dswap_k [--no-allow-shlib-undefined]
ld.lld: error: ../libopenblas_zen-r0.3.20.so: undefined reference to dcopy_k [--no-allow-shlib-undefined]
ld.lld: error: ../libopenblas_zen-r0.3.20.so: undefined reference to dasum_k [--no-allow-shlib-undefined]
ld.lld: error: ../libopenblas_zen-r0.3.20.so: undefined reference to dnrm2_k [--no-allow-shlib-undefined]
ld.lld: error: ../libopenblas_zen-r0.3.20.so: undefined reference to dmax_k [--no-allow-shlib-undefined]
ld.lld: error: ../libopenblas_zen-r0.3.20.so: undefined reference to damax_k [--no-allow-shlib-undefined]
ld.lld: error: too many errors emitted, stopping now (use -error-limit=0 to see all errors)
...
```

> new Makefile.system with the `-hnopattern` integrated [Makefile.system.txt](https://github.com/xianyi/OpenBLAS/files/9253362/Makefile.system.txt)

I modified the OpenBLAS Spack package to patch `Makefile.system` so that it reflects the changes you posted.
* https://github.com/eugeneswalker/spack/commit/aa7205b57bee345bf65d4dee71698409e03b1445
* https://github.com/eugeneswalker/spack/tree/openblas-cce-hnopattern-patch

I still see this error on Crusher using this patched branch:
```
$> spack install openblas%cce
...
==> Installing openblas-0.3.20-obtrraovvypinn3e3in2wqubpx2fewit
==> No binary for openblas-0.3.20-obtrraovvypinn3e3in2wqubpx2fewit found: installing from source
==> Fetching https://mirror.spack.io/_source-cache/archive/84/8495c9affc536253648e942908e88e097f2ec7753ede55aca52e5dead3029e3c.tar.gz
==> Applied patch /autofs/nccs-svm1_home1/lpeyrala/openblas-cray/spack/var/spack/repos/builtin/packages/openblas/cce-hnopattern.patch
==> openblas: Executing phase: 'edit'
==> openblas: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16' 'CC=/autofs/nccs-svm1_home1/lpeyrala/openblas-cray/spack/lib/spack/env/cce/cc' 'FC=/autofs/nccs-svm1_home1/lpeyrala/openblas-cray/spack/lib/spack/env/cce/ftn' 'MAKE_NB_JOBS=0' 'ARCH=x86_64' 'TARGET=ZEN' 'USE_LOCKING=1' 'USE_OPENMP=0' 'USE_THREAD=0' 'RANLIB=ranlib' 'libs' 'netlib' 'shared'

18 errors found in build log:
     2321    ftn-2307 ftn: ERROR in command line
     2322      The "-m" option must be followed by 0, 1, 2, 3 or 4.
     2323    ftn-2103 ftn: WARNING in command line
     2324      The -W all option is not supported or invalid and will be ignored.
     2325    ftn-2307 ftn: ERROR in command line
     2326      The "-m" option must be followed by 0, 1, 2, 3 or 4.
  >> 2327    make[2]: *** [<builtin>: sgbbrd.o] Error 1
     2328    make[2]: *** Waiting for unfinished jobs....
  >> 2329    make[2]: *** [<builtin>: sgbsv.o] Error 1
     2330    ftn-2103 ftn: WARNING in command line
     2331      The -W all option is not supported or invalid and will be ignored.
     2332    ftn-2307 ftn: ERROR in command line
     2333      The "-m" option must be followed by 0, 1, 2, 3 or 4.
     2334    ftn-2103 ftn: WARNING in command line
     2335      The -W all option is not supported or invalid and will be ignored.
  >> 2336    make[2]: *** [<builtin>: sgecon.o] Error 1
     2337    ftn-2307 ftn: ERROR in command line
     2338      The "-m" option must be followed by 0, 1, 2, 3 or 4.
     2339    ftn-2103 ftn: WARNING in command line
     2340      The -W all option is not supported or invalid and will be ignored.
     2341    ftn-2307 ftn: ERROR in command line
     2342      The "-m" option must be followed by 0, 1, 2, 3 or 4.
  >> 2343    make[2]: *** [<builtin>: sgbcon.o] Error 1
  >> 2344    make[2]: *** [<builtin>: sgebrd.o] Error 1
  >> 2345    make[2]: *** [<builtin>: sgbrfs.o] Error 1
  >> 2346    make[2]: *** [<builtin>: sgbsvx.o] Error 1
  >> 2347    make[2]: *** [<builtin>: sgbequ.o] Error 1
  >> 2348    make[2]: *** [<builtin>: sgbtf2.o] Error 1
  >> 2349    make[2]: *** [<builtin>: sgbtrf.o] Error 1
  >> 2350    make[2]: *** [<builtin>: sgetrf2.o] Error 1
  >> 2351    make[2]: *** [<builtin>: spotrf2.o] Error 1
  >> 2352    make[2]: *** [<builtin>: sgebd2.o] Error 1
  >> 2353    make[2]: *** [<builtin>: sgebak.o] Error 1
  >> 2354    make[2]: *** [<builtin>: sgbtrs.o] Error 1
  >> 2355    make[2]: *** [<builtin>: sgebal.o] Error 1
     2356    make[2]: Leaving directory '/tmp/lpeyrala/spack-stage/spack-stage-openblas-0.3.20-obtrraovvypinn3e3in2wqubpx2fewit/spack-src/lapack-netlib/SR
             C'
  >> 2357    make[1]: *** [Makefile:27: lapacklib] Error 2
     2358    make[1]: Leaving directory '/tmp/lpeyrala/spack-stage/spack-stage-openblas-0.3.20-obtrraovvypinn3e3in2wqubpx2fewit/spack-src/lapack-netlib'
  >> 2359    make: *** [Makefile:250: netlib] Error 2
```

Is this because `f_check` is needed in addition to the new `Makefile.system` ?

> PR updated now, should fix the `-m option should be followed by...` errors

Actually, I apologize, I had modified my Spack compiler entry to explicitly put `FFLAGS=-hnopattern` when I ran this. So the success was based on both your PR *and* setting `FFLAGS=-hnopattern` explicitly.

Using only your PR without the FFLAGS= addition to Spack's compiler definition does not work:
```
==> Installing openblas-0.3.20-hh3kja2miv5jumoe3kbrnume63tyv4yr
==> No binary for openblas-0.3.20-hh3kja2miv5jumoe3kbrnume63tyv4yr found: installing from source
==> No patches needed for openblas
==> openblas: Executing phase: 'edit'
==> openblas: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16' 'CC=/autofs/nccs-svm1_home1/lpeyrala/openblas-cray/spack/lib/spack/env/cce/cc' 'FC=/autofs/nccs-svm1_home1/lpeyrala/openblas-cray/spack/lib/spack/env/cce/ftn' 'MAKE_NB_JOBS=0' 'ARCH=x86_64' 'TARGET=ZEN' 'USE_LOCKING=1' 'USE_OPENMP=0' 'USE_THREAD=0' 'RANLIB=ranlib' 'libs' 'netlib' 'shared'

7 errors found in build log:
     6566    make[1]: Entering directory '/tmp/lpeyrala/spack-stage/spack-stage-openblas-0.3.20-hh3kja2miv5jumoe3kbrnume63tyv4yr/spack-src/exports'
     6567    ./gensymbol linktest  x86_64 _ 0 0 0 0 0 0 "" "" 1 0 1 1 1 1 > linktest.c
     6568    /autofs/nccs-svm1_home1/lpeyrala/openblas-cray/spack/lib/spack/env/cce/cc -O2 -DSMALL_MATRIX_OPT -DMAX_STACK_ALLOC=2048 -DUSE_LOCKING -Wall -m64 -fPIC -DNO_WARMUP -DMAX_CPU
             _NUMBER=128 -DMAX_PARALLEL_NUMBER=1 -DBUILD_SINGLE=1 -DBUILD_DOUBLE=1 -DBUILD_COMPLEX=1 -DBUILD_COMPLEX16=1 -DVERSION=\"0.3.20\" -msse3 -msse4.1 -mavx -mavx2 -mavx2 -UASMNA
             ME -UASMFNAME -UNAME -UCNAME -UCHAR_NAME -UCHAR_CNAME -DASMNAME= -DASMFNAME=_ -DNAME=_ -DCNAME= -DCHAR_NAME=\"_\" -DCHAR_CNAME=\"\" -DNO_AFFINITY -I..  -shared -o ../libope
             nblas_zen-r0.3.20.so \
     6569    -Wl,--whole-archive ../libopenblas_zen-r0.3.20.a -Wl,--no-whole-archive \
     6570    -Wl,-soname,libopenblas.so.0 -lm -lm
     6571    /autofs/nccs-svm1_home1/lpeyrala/openblas-cray/spack/lib/spack/env/cce/cc -O2 -DSMALL_MATRIX_OPT -DMAX_STACK_ALLOC=2048 -DUSE_LOCKING -Wall -m64 -fPIC -DNO_WARMUP -DMAX_CPU
             _NUMBER=128 -DMAX_PARALLEL_NUMBER=1 -DBUILD_SINGLE=1 -DBUILD_DOUBLE=1 -DBUILD_COMPLEX=1 -DBUILD_COMPLEX16=1 -DVERSION=\"0.3.20\" -msse3 -msse4.1 -mavx -mavx2 -mavx2 -UASMNA
             ME -UASMFNAME -UNAME -UCNAME -UCHAR_NAME -UCHAR_CNAME -DASMNAME= -DASMFNAME=_ -DNAME=_ -DCNAME= -DCHAR_NAME=\"_\" -DCHAR_CNAME=\"\" -DNO_AFFINITY -I..  -w -o linktest linkt
             est.c ../libopenblas_zen-r0.3.20.so -L/opt/cray/pe/gcc/10.3.0/snos/lib64 -L/opt/cray/pe/dsmml/0.2.2/dsmml//lib -L/opt/cray/pe/cce/14.0.0/cce/x86_64/lib/pkgconfig/../ -L/opt
             /gcc/10.3.0/snos/lib64 -L/opt/cray/pe/gcc/10.3.0/snos/lib64 -L/opt/cray/pe/cce/14.0.0/cce/x86_64/lib/pkgconfig/../ -L/opt/cray/pe/cce/14.0.0/cce-clang/x86_64/lib/clang/14.0
             .0/lib/linux -L/opt/cray/pe/gcc/10.3.0/snos/lib/gcc/x86_64-suse-linux/10.3.0 -L/opt/cray/pe/cce/14.0.0/binutils/x86_64/x86_64-pc-linux-gnu/..//x86_64-unknown-linux-gnu/lib
              -ldsmml -lstdc++ -lpgas-shmem -lquadmath -lmodules -lfi -lcraymath -lf -lu -lcsup -lgfortran -ltcmalloc_minimal -lstdc++ -lpthread -lc -lcsup -lm  -usystem_mod=/opt/cray/p
             e/cce/14.0.0/cce/x86_64/lib/libmodules.a -usystem_mod=/opt/cray/pe/cce/14.0.0/cce/x86_64/lib/libomp.a -usystem_mod=/opt/cray/pe/cce/14.0.0/cce/x86_64/lib/omp_lib.a -usystem
             _mod=/opt/cray/pe/cce/14.0.0/cce/x86_64/lib/libopenacc.a -usystem_mod=/opt/cray/pe/cce/14.0.0/cce/x86_64/lib/libcraymp.a && echo OK.
  >> 6572    ld.lld: error: ../libopenblas_zen-r0.3.20.so: undefined reference to _ismin_ [--no-allow-shlib-undefined]
  >> 6573    ld.lld: error: ../libopenblas_zen-r0.3.20.so: undefined reference to _idmin_ [--no-allow-shlib-undefined]
  >> 6574    ld.lld: error: ../libopenblas_zen-r0.3.20.so: undefined reference to _ismax_ [--no-allow-shlib-undefined]
  >> 6575    ld.lld: error: ../libopenblas_zen-r0.3.20.so: undefined reference to _idmax_ [--no-allow-shlib-undefined]
  >> 6576    clang-14: error: linker command failed with exit code 1 (use -v to see invocation)
  >> 6577    make[1]: *** [Makefile:196: ../libopenblas_zen-r0.3.20.so] Error 1
     6578    make[1]: Leaving directory '/tmp/lpeyrala/spack-stage/spack-stage-openblas-0.3.20-hh3kja2miv5jumoe3kbrnume63tyv4yr/spack-src/exports'
  >> 6579    make: *** [Makefile:125: shared] Error 2
```

To summarize:
* Explicit, manual setting of `FFLAGS=-hnopattern` *and* [PR3712](https://github.com/xianyi/OpenBLAS/pull/3712) = BUILDS OK
* [PR3712](https://github.com/xianyi/OpenBLAS/pull/3712) alone = BUILD FAILS

> Yep, smack in the middle of the OPEN64 conditional that was wider than it looked. Should be fixed now, sorry.

I think we're close! Still having the issue with `-m option must be followed by...`:

```
==> Installing openblas-0.3.20-hh3kja2miv5jumoe3kbrnume63tyv4yr
==> No binary for openblas-0.3.20-hh3kja2miv5jumoe3kbrnume63tyv4yr found: installing from source
==> No patches needed for openblas
==> openblas: Executing phase: 'edit'
==> openblas: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16' 'CC=/autofs/nccs-svm1_home1/lpeyrala/openblas-cray/spack/lib/spack/env/cce/cc' 'FC=/autofs/nccs-svm1_home1/lpeyrala/openblas-cray/spack/lib/spack/env/cce/ftn' 'MAKE_NB_JOBS=0' 'ARCH=x86_64' 'TARGET=ZEN' 'USE_LOCKING=1' 'USE_OPENMP=0' 'USE_THREAD=0' 'RANLIB=ranlib' 'libs' 'netlib' 'shared'

18 errors found in build log:
     2067    ftn-2307 ftn: ERROR in command line
     2068      The "-m" option must be followed by 0, 1, 2, 3 or 4.
     2069    ftn-2307 ftn: ERROR in command line
     2070      The "-m" option must be followed by 0, 1, 2, 3 or 4.
     2071    ftn-2307 ftn: ERROR in command line
     2072      The "-m" option must be followed by 0, 1, 2, 3 or 4.
  >> 2073    make[2]: *** [Makefile:64: spotrf2.o] Error 1
     2074    make[2]: *** Waiting for unfinished jobs....
  >> 2075    make[2]: *** [Makefile:64: sgebal.o] Error 1
  >> 2076    make[2]: *** [Makefile:64: sgbbrd.o] Error 1
  >> 2077    make[2]: *** [Makefile:64: sgbtrs.o] Error 1
  >> 2078    make[2]: *** [Makefile:64: sgbequ.o] Error 1
  >> 2079    make[2]: *** [Makefile:64: sgebak.o] Error 1
  >> 2080    make[2]: *** [Makefile:64: sgbtrf.o] Error 1
  >> 2081    make[2]: *** [Makefile:64: sgbcon.o] Error 1
  >> 2082    make[2]: *** [Makefile:64: sgebd2.o] Error 1
  >> 2083    make[2]: *** [Makefile:64: sgbsvx.o] Error 1
  >> 2084    make[2]: *** [Makefile:64: sgetrf2.o] Error 1
  >> 2085    make[2]: *** [Makefile:64: sgebrd.o] Error 1
  >> 2086    make[2]: *** [Makefile:64: sgecon.o] Error 1
  >> 2087    make[2]: *** [Makefile:64: sgbtf2.o] Error 1
  >> 2088    make[2]: *** [Makefile:64: sgbrfs.o] Error 1
  >> 2089    make[2]: *** [Makefile:64: sgbsv.o] Error 1
     2090    make[2]: Leaving directory '/tmp/lpeyrala/spack-stage/spack-stage-openblas-0.3.20-hh3kja2miv5jumoe3kbrnume63tyv4yr/spack-src/lapack-netlib/SRC'
  >> 2091    make[1]: *** [Makefile:27: lapacklib] Error 2
     2092    make[1]: Leaving directory '/tmp/lpeyrala/spack-stage/spack-stage-openblas-0.3.20-hh3kja2miv5jumoe3kbrnume63tyv4yr/spack-src/lapack-netlib'
  >> 2093    make: *** [Makefile:253: netlib] Error 2
```

But good news is `-hnopattern` is making it !
```
$> grep nopattern /tmp/lpeyrala/spack-stage/spack-stage-openblas-0.3.20-hh3kja2miv5jumoe3kbrnume63tyv4yr/spack-build-out.txt
/autofs/nccs-svm1_home1/lpeyrala/openblas-cray/spack/lib/spack/env/cce/ftn -O2 -hnopattern -O noomp -fPIC -msse3 -msse4.1 -mavx -mavx2 -c -o spotrf2.o spotrf2.f
/autofs/nccs-svm1_home1/lpeyrala/openblas-cray/spack/lib/spack/env/cce/ftn -O2 -hnopattern -O noomp -fPIC -msse3 -msse4.1 -mavx -mavx2 -c -o sgetrf2.o sgetrf2.f
/autofs/nccs-svm1_home1/lpeyrala/openblas-cray/spack/lib/spack/env/cce/ftn -O2 -hnopattern -O noomp -fPIC -msse3 -msse4.1 -mavx -mavx2 -c -o sgbbrd.o sgbbrd.f
/autofs/nccs-svm1_home1/lpeyrala/openblas-cray/spack/lib/spack/env/cce/ftn -O2 -hnopattern -O noomp -fPIC -msse3 -msse4.1 -mavx -mavx2 -c -o sgbcon.o sgbcon.f
/autofs/nccs-svm1_home1/lpeyrala/openblas-cray/spack/lib/spack/env/cce/ftn -O2 -hnopattern -O noomp -fPIC -msse3 -msse4.1 -mavx -mavx2 -c -o sgbequ.o sgbequ.f
/autofs/nccs-svm1_home1/lpeyrala/openblas-cray/spack/lib/spack/env/cce/ftn -O2 -hnopattern -O noomp -fPIC -msse3 -msse4.1 -mavx -mavx2 -c -o sgbrfs.o sgbrfs.f
/autofs/nccs-svm1_home1/lpeyrala/openblas-cray/spack/lib/spack/env/cce/ftn -O2 -hnopattern -O noomp -fPIC -msse3 -msse4.1 -mavx -mavx2 -c -o sgbsv.o sgbsv.f
/autofs/nccs-svm1_home1/lpeyrala/openblas-cray/spack/lib/spack/env/cce/ftn -O2 -hnopattern -O noomp -fPIC -msse3 -msse4.1 -mavx -mavx2 -c -o sgbsvx.o sgbsvx.f
/autofs/nccs-svm1_home1/lpeyrala/openblas-cray/spack/lib/spack/env/cce/ftn -O2 -hnopattern -O noomp -fPIC -msse3 -msse4.1 -mavx -mavx2 -c -o sgbtf2.o sgbtf2.f
/autofs/nccs-svm1_home1/lpeyrala/openblas-cray/spack/lib/spack/env/cce/ftn -O2 -hnopattern -O noomp -fPIC -msse3 -msse4.1 -mavx -mavx2 -c -o sgbtrf.o sgbtrf.f
/autofs/nccs-svm1_home1/lpeyrala/openblas-cray/spack/lib/spack/env/cce/ftn -O2 -hnopattern -O noomp -fPIC -msse3 -msse4.1 -mavx -mavx2 -c -o sgbtrs.o sgbtrs.f
/autofs/nccs-svm1_home1/lpeyrala/openblas-cray/spack/lib/spack/env/cce/ftn -O2 -hnopattern -O noomp -fPIC -msse3 -msse4.1 -mavx -mavx2 -c -o sgebak.o sgebak.f
/autofs/nccs-svm1_home1/lpeyrala/openblas-cray/spack/lib/spack/env/cce/ftn -O2 -hnopattern -O noomp -fPIC -msse3 -msse4.1 -mavx -mavx2 -c -o sgebal.o sgebal.f
/autofs/nccs-svm1_home1/lpeyrala/openblas-cray/spack/lib/spack/env/cce/ftn -O2 -hnopattern -O noomp -fPIC -msse3 -msse4.1 -mavx -mavx2 -c -o sgebd2.o sgebd2.f
/autofs/nccs-svm1_home1/lpeyrala/openblas-cray/spack/lib/spack/env/cce/ftn -O2 -hnopattern -O noomp -fPIC -msse3 -msse4.1 -mavx -mavx2 -c -o sgebrd.o sgebrd.f
/autofs/nccs-svm1_home1/lpeyrala/openblas-cray/spack/lib/spack/env/cce/ftn -O2 -hnopattern -O noomp -fPIC -msse3 -msse4.1 -mavx -mavx2 -c -o sgecon.o sgecon.f
```

so we are close!

--- a/Makefile.system
+++ b/Makefile.system
@@ -1580,6 +1580,10 @@ ifeq ($(F_COMPILER),NAG)
 LAPACK_FFLAGS := $(filter-out -msse3 -mssse3 -msse4.1 -mavx -mavx2 -mskylake-avx512 ,$(FFLAGS))
 endif
 
+ifeq ($(F_COMPILER),CRAY)
+LAPACK_FFLAGS := $(filter-out -msse3 -mssse3 -msse4.1 -mavx -mavx2 -mskylake-avx512 ,$(FFLAGS))
+endif
+
 LAPACK_CFLAGS = $(CFLAGS)
 LAPACK_CFLAGS += -DHAVE_LAPACK_CONFIG_H
 ifdef INTERFACE64

--- a/Makefile.system
+++ b/Makefile.system
@@ -263,6 +263,7 @@ ifeq ($(ARCH), x86_64)
 SMALL_MATRIX_OPT = 1
 else ifeq ($(ARCH), power)
 SMALL_MATRIX_OPT = 1
+BUILD_BFLOAT16 = 1
 endif
 ifeq ($(SMALL_MATRIX_OPT), 1)
 CCOMMON_OPT += -DSMALL_MATRIX_OPT
@@ -677,6 +678,10 @@ endif
 
 ifeq ($(ARCH), mips64)
 DYNAMIC_CORE = LOONGSON3R3 LOONGSON3R4
+endif
+
+ifeq ($(ARCH), loongarch64)
+DYNAMIC_CORE = LOONGSON3R5 LOONGSON2K1000 LOONGSONGENERIC
 endif
 
 ifeq ($(ARCH), zarch)
@@ -1570,7 +1575,6 @@ endif
 ifeq ($(F_COMPILER),CRAY)
 LAPACK_FFLAGS := $(filter-out -msse3 -mssse3 -msse4.1 -mavx -mavx2 -mskylake-avx512 ,$(FFLAGS))
 FFLAGS := $(filter-out -msse3 -mssse3 -msse4.1 -mavx -mavx2 -mskylake-avx512 ,$(FFLAGS))
-endif
 endif
 
 LAPACK_CFLAGS = $(CFLAGS)

--- a/Makefile.system
+++ b/Makefile.system
@@ -1565,6 +1565,12 @@ endif
 
 ifeq ($(F_COMPILER),NAG)
 LAPACK_FFLAGS := $(filter-out -msse3 -mssse3 -msse4.1 -mavx -mavx2 -mskylake-avx512 ,$(FFLAGS))
+FFLAGS := $(filter-out -msse3 -mssse3 -msse4.1 -mavx -mavx2 -mskylake-avx512 ,$(FFLAGS))
+endif
+ifeq ($(F_COMPILER),CRAY)
+LAPACK_FFLAGS := $(filter-out -msse3 -mssse3 -msse4.1 -mavx -mavx2 -mskylake-avx512 ,$(FFLAGS))
+FFLAGS := $(filter-out -msse3 -mssse3 -msse4.1 -mavx -mavx2 -mskylake-avx512 ,$(FFLAGS))
+endif
 endif
 
 LAPACK_CFLAGS = $(CFLAGS)

--- a/Makefile.system
+++ b/Makefile.system
@@ -1191,6 +1191,19 @@ endif
 endif
 endif
 
+ifeq ($(F_COMPILER), CRAY)
+CCOMMON_OPT += -DF_INTERFACE_INTEL
+FCOMMON_OPT += -hnopattern
+ifdef INTERFACE64
+ifneq ($(INTERFACE64), 0)
+FCOMMON_OPT += -s integer64
+endif
+endif
+ifneq ($(USE_OPENMP), 1)
+FCOMMON_OPT += -O noomp
+endif
+endif
+
 ifeq ($(ARCH), $(filter $(ARCH),mips64 mips))
 ifndef BINARY64
 FCOMMON_OPT += -n32

--- a/Makefile.system
+++ b/Makefile.system
@@ -263,7 +263,6 @@ ifeq ($(ARCH), x86_64)
 SMALL_MATRIX_OPT = 1
 else ifeq ($(ARCH), power)
 SMALL_MATRIX_OPT = 1
-BUILD_BFLOAT16 = 1
 endif
 ifeq ($(SMALL_MATRIX_OPT), 1)
 CCOMMON_OPT += -DSMALL_MATRIX_OPT
@@ -678,10 +677,6 @@ endif
 
 ifeq ($(ARCH), mips64)
 DYNAMIC_CORE = LOONGSON3R3 LOONGSON3R4
-endif
-
-ifeq ($(ARCH), loongarch64)
-DYNAMIC_CORE = LOONGSON3R5 LOONGSON2K1000 LOONGSONGENERIC
 endif
 
 ifeq ($(ARCH), zarch)
@@ -1190,20 +1185,6 @@ FCOMMON_OPT += -i8
 endif
 endif
 endif
-
-ifeq ($(F_COMPILER), CRAY)
-CCOMMON_OPT += -DF_INTERFACE_INTEL
-FCOMMON_OPT += -hnopattern
-ifdef INTERFACE64
-ifneq ($(INTERFACE64), 0)
-FCOMMON_OPT += -s integer64
-endif
-endif
-ifneq ($(USE_OPENMP), 1)
-FCOMMON_OPT += -O noomp
-endif
-endif
-
 ifeq ($(ARCH), $(filter $(ARCH),mips64 mips))
 ifndef BINARY64
 FCOMMON_OPT += -n32
@@ -1213,11 +1194,9 @@ endif
 ifeq ($(CORE), LOONGSON3R3)
 FCOMMON_OPT += -loongson3 -static
 endif
-
 ifeq ($(CORE), LOONGSON3R4)
 FCOMMON_OPT += -loongson3 -static
 endif
-
 else
 ifndef BINARY64
 FCOMMON_OPT += -m32
@@ -1225,7 +1204,6 @@ else
 FCOMMON_OPT += -m64
 endif
 endif
-
 ifeq ($(USE_OPENMP), 1)
 FEXTRALIB   += -lstdc++
 FCOMMON_OPT += -mp
@@ -1233,7 +1211,6 @@ endif
 endif
 
 ifeq ($(C_COMPILER), OPEN64)
-
 ifeq ($(ARCH), $(filter $(ARCH),mips64 mips))
 ifndef BINARY64
 CCOMMON_OPT += -n32
@@ -1243,13 +1220,10 @@ endif
 ifeq ($(CORE), LOONGSON3R3)
 CCOMMON_OPT += -loongson3 -static
 endif
-
 ifeq ($(CORE), LOONGSON3R4)
 CCOMMON_OPT += -loongson3 -static
 endif
-
 else
-
 ifndef BINARY64
 CCOMMON_OPT += -m32
 else
@@ -1292,6 +1266,19 @@ ifeq ($(F_COMPILER), COMPAQ)
 CCOMMON_OPT  += -DF_INTERFACE_COMPAQ
 ifeq ($(USE_OPENMP), 1)
 FCOMMON_OPT += -openmp
+endif
+endif
+
+ifeq ($(F_COMPILER), CRAY)
+CCOMMON_OPT += -DF_INTERFACE_INTEL
+FCOMMON_OPT += -hnopattern
+ifdef INTERFACE64
+ifneq ($(INTERFACE64), 0)
+FCOMMON_OPT += -s integer64
+endif
+endif
+ifneq ($(USE_OPENMP), 1)
+FCOMMON_OPT += -O noomp
 endif
 endif
 
@@ -1577,10 +1564,6 @@ LAPACK_FPFLAGS := $(FPFLAGS)
 endif
 
 ifeq ($(F_COMPILER),NAG)
-LAPACK_FFLAGS := $(filter-out -msse3 -mssse3 -msse4.1 -mavx -mavx2 -mskylake-avx512 ,$(FFLAGS))
-endif
-
-ifeq ($(F_COMPILER),CRAY)
 LAPACK_FFLAGS := $(filter-out -msse3 -mssse3 -msse4.1 -mavx -mavx2 -mskylake-avx512 ,$(FFLAGS))
 endif
 

--- a/cmake/fc.cmake
+++ b/cmake/fc.cmake
@@ -222,6 +222,17 @@ if (${F_COMPILER} STREQUAL "COMPAQ")
   endif ()
 endif ()
 
+if (${F_COMPILER} STREQUAL "CRAY")
+  set(CCOMMON_OPT "${CCOMMON_OPT} -DF_INTERFACE_INTEL")
+  set(FCOMMON_OPT "${FCOMMON_OPT} -hnopattern")
+  if (INTERFACE64)
+    set (FCOMMON_OPT "${FCOMMON_OPT} -s integer64")
+  endif ()
+  if (NOT USE_OPENMP)
+    set(FCOMMON_OPT "${FCOMMON_OPT} -O noomp")
+  endif ()
+endif ()
+
 # from the root Makefile - this is for lapack-netlib to compile the correct secnd file.
 if (${F_COMPILER} STREQUAL "GFORTRAN")
   set(TIMER "INT_ETIME")

--- a/cmake/system.cmake
+++ b/cmake/system.cmake
@@ -559,6 +559,14 @@ if (${CMAKE_SYSTEM_NAME} STREQUAL "Windows")
   endforeach ()
 endif ()
 
+if ("${F_COMPILER}" STREQUAL "NAG" OR "${F_COMPILER}" STREQUAL "CRAY")
+  set(FILTER_FLAGS "-msse3;-mssse3;-msse4.1;-mavx;-mavx2,-mskylake-avx512")
+  foreach (FILTER_FLAG ${FILTER_FLAGS})
+    string(REPLACE ${FILTER_FLAG} "" LAPACK_FFLAGS ${LAPACK_FFLAGS})
+    string(REPLACE ${FILTER_FLAG} "" LAPACK_FPFLAGS ${LAPACK_FPFLAGS})
+  endforeach ()
+endif ()
+
 if ("${F_COMPILER}" STREQUAL "GFORTRAN")
   # lapack-netlib is rife with uninitialized warnings -hpa
   set(LAPACK_FFLAGS "${LAPACK_FFLAGS} -Wno-maybe-uninitialized")

--- a/f_check
+++ b/f_check
@@ -299,6 +299,9 @@ fi
 if [ "$vendor" = "NAG" ]; then
     link=`$compiler $openmp -dryrun ftest2.f 2>&1 && rm -f a.out a.exe`
 fi
+if [ "$vendor" = "CRAY" ]; then
+    link=`$compiler $openmp -hnopattern ftest2.f 2>&1 && rm -f a.out a.exe`
+fi
 linker_L=""
 linker_l=""
 linker_a=""

--- a/f_check
+++ b/f_check
@@ -45,7 +45,7 @@ if [ -z "$compiler" ]; then
 	      pathf90 pathf95
 	      pgf95 pgf90 pgf77 pgfortran nvfortran
 	      flang egfortran
-              ifort nagfor ifx"
+              ifort nagfor ifx ftn crayftn"
 
     for list in $lists; do
         for p in $path; do
@@ -82,6 +82,10 @@ else
                 vendor=FUJITSU
                 openmp='-Kopenmp'
                 ;;
+	    *Cray*)
+		vendor=CRAY
+		openmp='-fopenmp'
+		;;
             *GNU*|*GCC*)
 
                 v="${data#*GCC: *\) }"

--- a/f_check.pl
+++ b/f_check.pl
@@ -34,7 +34,7 @@ if ($compiler eq "") {
 	      "pathf90", "pathf95",
 	      "pgf95", "pgf90", "pgf77", "pgfortran", "nvfortran",
 	      "flang", "egfortran",
-              "ifort", "nagfor", "ifx");
+              "ifort", "nagfor", "ifx", "ftn", "crayftn");
 
 OUTER:
     foreach $lists (@lists) {
@@ -75,6 +75,11 @@ if ($compiler eq "") {
 
 	    $vendor = FUJITSU;
 	    $openmp = "-Kopenmp";
+
+	} elsif ($data =~ /Cray/) {
+
+	    $vendor = CRAY;
+	    $openmp = "-fopenmp";
 
 	} elsif ($data =~ /GNU/ || $data =~ /GCC/ ) {
 

--- a/f_check.pl
+++ b/f_check.pl
@@ -314,6 +314,9 @@ if (!$?) {
 if ( $vendor eq "NAG") {
 	    $link = `$compiler $openmp -dryrun ftest2.f 2>&1 && rm -f a.out a.exe`;
     }
+if ( $vendor eq "CRAY") {
+	    $link = `$compiler $openmp -hnopattern ftest2.f 2>&1 && rm -f a.out a.exe`;
+    }
 $linker_L = "";
 $linker_l = "";
 $linker_a = "";

